### PR TITLE
ConversationEntityMemory: Add multiple prompts and examples

### PR DESCRIPTION
**Pull request by sjwhitmore**

### CODE REVIEW

File: langchain/docstore/in_memory.py [hunk 1]
```diff
@@ -12,10 +12,11 @@ def __init__(self, _dict: Dict[str, Document]):
         """Initialize with dict."""
         self._dict = _dict
 
-    def add(self, texts: Dict[str, Document]) -> None:
+    def add(self, texts: Dict[str, Document], **kwargs: Any) -> None:
```
Comment (hwchase17) [inline langchain/docstore/in_memory.py:15]: hm i dont love this. will other docstores haave this?

File: langchain/chains/conversation/memory.py [hunk 2]
```diff
@@ -40,6 +46,88 @@ def _save_context(self, inputs: Dict[str, Any], outputs: Dict[str, str]) -> None
...
+    @property
+    def dynamic_keys(self) -> List[str]:
+        """Will always return list of dynamic keys.
+
+        :meta private:
+        """
+        return self.dynamic_key_list
+
+    def _format_buffer_to_string(self) -> str:
+        section = self.buffer[-self.history_window_len :]
```
Comment (hwchase17) [inline langchain/chains/conversation/memory.py:70]: will we really want this on the other memory objects as well?

File: langchain/chains/conversation/memory.py [hunk 3]
```diff
@@ -40,6 +46,88 @@ def _save_context(self, inputs: Dict[str, Any], outputs: Dict[str, str]) -> None
...
+        result = []
+        for line in section:
+            result.append("Human: {input}".format(input=line["input"]))
+            result.append("AI: {response}".format(response=line["response"]))
+        return "\n".join(result)
+
+    def _load_dynamic_keys(self, inputs: Dict[str, Any]) -> Dict[str, str]:
+        """Return history buffer with entity summaries."""
+        prompt_input_keys = list(set(inputs.keys()) - set(self.dynamic_keys))
+        assert len(prompt_input_keys) == 1, "One input key expected"
```
Comment (hwchase17) [inline langchain/chains/conversation/memory.py:80]: nit: raise ValueError instead of assert

File: langchain/chains/conversation/memory.py [hunk 4]
```diff
@@ -40,6 +46,88 @@ def _save_context(self, inputs: Dict[str, Any], outputs: Dict[str, str]) -> None
...
+        section = self.buffer[-self.history_window_len :]
+        result = []
+        for line in section:
+            result.append("Human: {input}".format(input=line["input"]))
+            result.append("AI: {response}".format(response=line["response"]))
+        return "\n".join(result)
+
+    def _load_dynamic_keys(self, inputs: Dict[str, Any]) -> Dict[str, str]:
+        """Return history buffer with entity summaries."""
+        prompt_input_keys = list(set(inputs.keys()) - set(self.dynamic_keys))
```
Comment (hwchase17) [inline langchain/chains/conversation/memory.py:79]: nit: list(set(inputs).difference(self.dynamic_kays)))

File: langchain/chains/conversation/memory.py [hunk 5]
```diff
@@ -40,6 +46,88 @@ def _save_context(self, inputs: Dict[str, Any], outputs: Dict[str, str]) -> None
...
+        assert len(prompt_input_keys) == 1, "One input key expected"
+        input = inputs[prompt_input_keys[0]]
+        full_conversation = self._format_buffer_to_string() + "\nHuman: " + input
```
Comment (hwchase17) [inline langchain/chains/conversation/memory.py:82]: as a tech debt thing, everything where we do string manipulation here and in other memorty classes we should try to make a prompt template. cn you add a bit todo at top of file?

File: langchain/chains/conversation/memory.py [hunk 6]
```diff
@@ -40,6 +46,88 @@ def _save_context(self, inputs: Dict[str, Any], outputs: Dict[str, str]) -> None
...
+                all_results += "\nHere's what we know about {entity}".format(
+                    entity=entity
+                )
+                all_results += "\n" + entity_summary.page_content
+            self.entities_per_turn = all_entities
+        return {"entities": all_results, "history": self._format_buffer_to_string()}
+
+    def _save_context(self, inputs: Dict[str, Any], outputs: Dict[str, str]) -> None:
+        """Save context from this conversation to buffer."""
+        prompt_input_keys = list(set(inputs.keys()) - set(self.dynamic_keys))
```
Comment (hwchase17) [inline langchain/chains/conversation/memory.py:106]: same nits as above re set(inputs) and raise valuerror

File: langchain/chains/conversation/memory.py [hunk 7]
```diff
@@ -40,6 +46,88 @@ def _save_context(self, inputs: Dict[str, Any], outputs: Dict[str, str]) -> None
...
+            "entity",
+            "input",
+            "summary",
+        ], "Expected different input keys"
+        human = inputs[prompt_input_keys[0]]
+        ai = outputs[list(outputs.keys())[0]]
+        full_conversation = self._format_buffer_to_string() + "\nHuman: " + human
+        for entity in self.entities_per_turn:
+            chain = LLMChain(llm=self.llm, prompt=self.entity_summarization_prompt)
+            entity_summary = self.docstore.search(entity) or ""
```
Comment (hwchase17) [inline langchain/chains/conversation/memory.py:120]: wait doesnt search return a document object?

File: langchain/chains/conversation/memory.py (hunk 2 above)
Comment (hwchase17) [reply langchain/chains/conversation/memory.py:70]: if so, i wonder if theres a way to use two memory objects (as far as i cn tell, the history and entity logic is pretty unrelated?

File: langchain/docstore/in_memory.py (hunk 1 above)
Comment (sjwhitmore) [reply langchain/docstore/in_memory.py:15]: i do think its a useful abstraction for docstores -- I overwrite key value stores all the time if there's new information about the key. I could separate it out into a different PR & update the base class too if that makes more sense?

File: langchain/chains/conversation/memory.py (hunk 7 above)
Comment (sjwhitmore) [reply langchain/chains/conversation/memory.py:120]: InMemoryDocstore at the moment returns a string response if the document isnt found (None might be better tbh)

**Pull request closed without merging**